databases/week2/lesson2.sql

```diff
@@ -0,0 +1,228 @@
+--Class: with the columns: id, name, begins (date), ends (date)
+--Student: with the columns: id, name, email, phone, class_id (foreign key)
+CREATE database School;
+use School;
+create table `Class`(
+`id` INT (10) UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+`name` VARCHAR (255) NOT NULL,
+`begins` DATETIME,
+`ends` DATETIME 
+);
+CREATE Table `student`(
+`id` INT (10) UNSIGNED PRIMARY KEY AUTO_INCREMENT,
+`name` VARCHAR (255) NOT NULL,
+`email` VARCHAR (255) NOT NULL,
+`phone` VARCHAR(255) NULL,
+`class_id` INT (10) UNSIGNED NOT NULL,
+FOREIGN KEY (`Class_id`) REFERENCES Class(`id`) ON delete SET NULL ON update CASCADE
+);
+--Create an index on the name column of the student table.
+--add a new column to the class table named status which can only have 
+--the following values: not-started, ongoing, finished (hint: enumerations).
+CREATE index idx_name
+on student (name);
+SHOW INDEX FROM student;
+ALTER Table Class
+ADD status enum('not-started', 'ongoing', 'finished') not null;
+----------------------------------------------------------------------
+-------------------------Part 3: More queries
+use hyf_lesson2;
+
+
+--Get all the tasks assigned to users whose email ends in @spotify.com
+SELECT
+user.email,
+task.title,
+task.description
+FROM (task JOIN user_task ON task.id = user_task.task_id)
+JOIN user ON user.id = user_task.user_id
+where user.email like '%@spotify.com';
+
+--Get all the tasks for 'Donald Duck' with status 'Not started'
+SELECT
+user.name as username,
+task.title,
+status.name 
+FROM (task join user_task ON task.id = user_task.task_id) 
+JOIN user ON user.id = user_task.user_id
+join status ON status.id= task.status_id 
+where status.name = "Not started" AND user.name = 'Donald Duck';
+
+--Get all the tasks for 'Maryrose Meadows' that were created in september (hint: month(created)=month_number)
+SELECT
+user.name,
+task.title,
+task.created
+FROM (task JOIN user_task ON task.id = user_task.task_id)
+JOIN user ON user.id = user_task.user_id
+WHERE month(task.created) = 9 AND user.name = "Maryrose Meadows";
+
+--Find how many tasks where created in each month, e.g. 
+--how many tasks were created in october, how many tasks were created in 
+--november, etc. (hint: use group by)
+SELECT
+month(task.created) as months,
+COUNT(task.id) as task_count
+FROM task
+GROUP BY month(task.created);
+---------------------------------------------------------------------------
+
+--------------------------------------Part 4: Creating a database
+-- create a simple university register database
+use registration;
+SET NAMES utf8mb4;
+-- database has 4 tables: student, department, course and enrollment.
+CREATE Table `student`(
+    student_id INT(10) UNSIGNED PRIMARY KEY,
+    Name VARCHAR (255) NOT NULL,
+    GPA DECIMAL(3,2)
+);
+-- delete auto_crement from student_id
+--ALTER table student 
+--DROP PRIMARY KEY,
+--CHANGE student_id student_id INT(10) UNSIGNED PRIMARY KEY;
+CREATE Table `department`(
+    dept_id char(8) PRIMARY KEY,
```

Comment:
Maybe you can keep using the `int(10)` as a type for the id. 
Unless there is a specific reason why you chose to go with this?

Reply:
Thank you so much for great comments.
About char(8) for department.id, I choose this and in fact I saw this in net. I thought maybe as universities id can be a abbreviation for departments, f.x IE for industrial engineering. Is there any preferences between them? 
I removed accidently your first comment. sorry for that.

Reply:
I see, I think it might vary based on the uscase.
From my experience, tables always have an Id of a numeric type (int/long) and if there is a need to store other types of identifiers then there are separate columns. based on what you mentioned you could have an Id (int) and Abbreviation (nvarchar).
This would have multiple benefits:
1. the table would be standardized as it is a common practice to use int as ids.
2. Indexing the table would be more efficient
3. the int id can be generated automatically as mysql has out of the box functionality for this.
4. if a specific entry does not have an abbreviation then that column can be left empty without altering the consistency of Ids

If you have found other information please share as I am curious what other ideas are out there :) 

Let me know if you have any questions or if you would like me to share some articles or documentation about this 😁

Reply:
Thank you so much for taking time and clarifying so detail my question and also reviewing my files.:)
I actually did not search about it but if I find I will certainly send you.

Reply:
My pleasure! I am happy that you found it useful :)